Disable useless error-prone checks (libraries/frameworks we do not use)

rmuir opened the pull request:
These are easy/obvious ones to disable since we don't use the functionality at all: the checks are literally useless.

This gives some performance boost to the error-prone, although it is still pretty slow.

rmuir committed:
Disable useless error-prone checks (libraries/frameworks we do not use)

These are easy/obvious ones to disable since we don't use the
functionality at all: the checks are literally useless.

This gives some performance boost to the error-prone, although it is
still pretty slow.

rmuir commented:
Before:
```
 225.22 sec.  compileTestJava
 223.56 sec.  compileJava
```

After:
```
 158.53 sec.  compileJava
 153.20 sec.  compileTestJava
```

rmuir commented:
I've at least checked the disabled ones includes the ones @madrob pointed out in #11970 . But like I said, I went thru the checks so fast that i'm sure i missed a few we don't need.

madrob reviewed (commented):
Github won't let me comment on the line but we don't use theory library

madrob commented on gradle/validation/error-prone.gradle:
```diff
@@ -123,36 +123,36 @@ allprojects { prj ->
-            '-Xep:IncompatibleModifiers:ERROR',
+            // '-Xep:IncompatibleArgumentType:OFF', // we don't use this annotation
+            // '-Xep:IncompatibleModifiers:OFF', // we don't use this annotation
             '-Xep:IndexOfChar:ERROR',
```
I don't understand this check from reading the bug pattern description https://errorprone.info/bugpattern/IndexOfChar but maybe forbiddenAPIs can handle it?

rmuir replied:
Yeah, confirmed that's the case. See https://github.com/google/error-prone/blob/master/core/src/main/java/com/google/errorprone/bugpatterns/IndexOfChar.java#L58-L59

madrob commented on gradle/validation/error-prone.gradle:
```diff
@@ -161,65 +161,65 @@ allprojects { prj ->
             '-Xep:IsInstanceIncompatibleType:ERROR',
             '-Xep:IsInstanceOfClass:ERROR',
             '-Xep:IsLoggableTagLength:ERROR',
```
This should be forbidden APIs to prevent logging with arbitrary tags anyway.

rmuir replied:
our use of logging is extremely minimal (intentionally) in lucene. Will take another look at this check.

rmuir replied:
this is an android specific check. disabling.

madrob commented on gradle/validation/error-prone.gradle:
```diff
@@ -161,65 +161,65 @@ allprojects { prj ->
             '-Xep:NCopiesOfChar:ERROR',
-            '-Xep:NoCanIgnoreReturnValueOnClasses:ERROR',
+            // '-Xep:NoCanIgnoreReturnValueOnClasses:OFF', // we don't use this annotation
             '-Xep:NonCanonicalStaticImport:ERROR',
```
I don't think this is a worthwhile check regardless, but that's just my opinion

rmuir replied:
Yeah, a lot of checks like these maybe got "implicitly" enabled when we upgraded error-prone. All the new checks got enabled even though we didn't opt-in. Thats the problem with the old blocklist approach. So really we need to do a pass over all the checks and figure out which ones we want/don't want.

madrob commented on gradle/validation/error-prone.gradle:
```diff
@@ -256,30 +256,30 @@ allprojects { prj ->
             '-Xep:BadComparable:WARN',
             // '-Xep:BadImport:OFF',
             // '-Xep:BadInstanceof:OFF',
             '-Xep:BareDotMetacharacter:WARN',
```
I remember this coming up before. I don't remember if forbidden API or something else found it for us. Maybe it was error prone. I feel like Uwe was the one to fix it though...

madrob commented on gradle/validation/error-prone.gradle:
```diff
@@ -362,33 +362,33 @@ allprojects { prj ->
             // '-Xep:MathAbsoluteRandom:OFF',
-            '-Xep:MemoizeConstantVisitorStateLookups:WARN',
+            // '-Xep:MemoizeConstantVisitorStateLookups:OFF', // we don't use this class
             '-Xep:MissingCasesInEnumSwitch:WARN',
```
Does javac do this?

rmuir replied:
we turn this on in `ECJ` linter already, let's nuke it here.

madrob commented on gradle/validation/error-prone.gradle:
```diff
@@ -417,21 +417,21 @@ allprojects { prj ->
             '-Xep:OutlineNone:WARN',
```
This is for CSS web styles?

rmuir replied:
LOL will take a look again, wasn't sure whats happening here. Sometimes you have to look at source code to the checker since the description may not be great.

actually a lot of the javadocs checks may be redundant. We crank up javadocs checking with openjdk javadocs linting and some of our own checks (e.g. broken link detection and so on). will take another look here.

rmuir replied:
This check is looking for certain calls to google GWT, which we don't use: https://github.com/google/error-prone/blob/master/core/src/main/java/com/google/errorprone/bugpatterns/OutlineNone.java#L84-L87

madrob commented on gradle/validation/error-prone.gradle:
```diff
@@ -442,18 +442,18 @@ allprojects { prj ->
             // '-Xep:StreamResourceLeak:OFF',
             '-Xep:StreamToIterable:WARN',
             '-Xep:StringSplitter:OFF',
```
needs a comment

rmuir replied:
pre-existing condition: it was in the list of disabled ones originally with no comment: i preserved these as-is. each one should be evaluated as to why its disabled, and a comment added. probably many should actually be enabled...

madrob commented on gradle/validation/error-prone.gradle:
```diff
@@ -476,11 +476,11 @@ allprojects { prj ->
             // '-Xep:WaitNotInLoop:OFF',
-            '-Xep:WakelockReleasedDangerously:WARN',
+            // '-Xep:WakelockReleasedDangerously:OFF', // we don't use android
             '-Xep:WithSignatureDiscouraged:WARN',
```
I have no idea what this applies to but it's probably useless

rmuir replied:
https://github.com/google/error-prone/blob/master/core/src/main/java/com/google/errorprone/bugpatterns/WithSignatureDiscouraged.java#L50

Doesn't apply to us, seems only of use to error-prone itself. disabling this.

madrob commented:
> > Github won't let me comment on the line but we don't use theory library
> 
> Do you refer to this check? https://errorprone.info/bugpattern/TheoryButNoTheories
> 
> I didn't touch it simply because these classes (Theory/DataPoint/DataPoints) are in junit API. But yeah, if we don't do those kinds of tests, let's nuke this check.

Ah, ok, they used to be in a separate library but I guess JUnit adopted them. We don't have any of those annotations. And I think they require a specialized runner anyway, which might not be compatible with RandomizedRunner.

rmuir committed:
disable theoretical TheoriesButNoTheories since we don't use @Theory

rmuir commented:
thanks @madrob for taking a look. ill look deeper into the ones you brought up here.

rmuir committed:
disable error-prone missing-enum-case as ECJ already takes care of this

rmuir committed:
disable GWT style check since we don't use it

rmuir committed:
disable check only useful to error-prone developers...

rmuir committed:
disable android specific check

rmuir committed:
disable @Inject specific checks, we don't use it

rmuir committed:
disable check specific to mock/@Inject

rmuir committed:
disable checks for issues prevented by LuceneTestCase

rmuir committed:
ban JNDI with forbidden-apis instead of error-prone

rmuir committed:
disable checks against java time apis that we don't even use

rmuir commented:
Still working thru these, but the execution time is getting better:

```
 144.97 sec.  compileJava
 136.46 sec.  compileTestJava
```

Big improvement since around a week ago when it looked like this: (before #11927 etc):
```
 550.91 sec.  compileJava
 384.45 sec.  compileTestJava
```

Basically findings are lining up with behavior I see in https://github.com/google/error-prone/issues/994

rmuir committed:
disable a few more useless checks

rmuir committed:
remove more useless guava etc checks

rmuir commented:
Only about 2x as costly as `javac` now:
```
 129.49 sec.  compileJava
 122.41 sec.  compileTestJava
```

I need to take another pass thru each check. Many of these checks have a great-sounding name, description sounds wonderful, then you look at the source code to the check and see that it only works on google libraries :) So I keep finding more to disable...

rmuir committed:
triage most of the previously disabled checks into TODO, noisy, etc

Running them against the codebase, inspecting the source code, etc.

Quite a few of these are good ones to fix.

dweiss commented:
From that link:
> There are a couple of checks that -XepDisableAllChecks doesn't disable

LOL. Of course, why would it work as the name suggests... ;)

dweiss commented:
Excellent improvement, btw.

rmuir committed:
disable more google-library specific checks

rmuir committed:
disable more useless checks

rmuir commented:
I think its enough for this issue. Majority of the error-prone checks we had enabled were just slowing down the build and providing no benefit :(

And the majority of the ones that could provide benefit were disabled. But those require some work, for other issues.

rmuir commented:
i'm happy if we disable more of the currently-enabled checks in followup issues. It starts to become subjective, but honestly most of the checks we have enabled are useless.

If we haven't done explicit work to fix  the associated problems across our (large) codebase, then such static analysis checks probably aren't doing anything. 

As a next step I'd rather work on enabling some of the checks marked `TODO: there are problems` here. Those are ones that I tried out, and the failures look like possibly real bugs.

rmuir committed:
Merge branch 'main' into disable_useless_errorprone

rmuir committed:
document AlmostJavaDoc as noisy, since commented-out-code gets detected as comment-that-should-be-javadoc

rmuir commented:
no feedback for a few days, i'm going to merge this as it disables stuff that's really not controversial, as it is providing no value. as far as which of the other checks are enabled, we can debate in other issues.
